src/EditorFeatures/Core/Structure/AbstractStructureTaggerProvider.cs

```diff
@@ -67,37 +67,146 @@ protected override bool ComputeInitialTagsSynchronously(ITextBuffer subjectBuffe
-            if (isMetadataAsSource)
+            if (openDocument.Project.Solution.Workspace.Kind == WorkspaceKind.MetadataAsSource &&
+                GlobalOptions.GetOption(BlockStructureOptionsStorage.CollapseMetadataImplementationsWhenFirstOpened, openDocument.Project.Language))
+            {
                 return true;
+            }
+
+            // If the user wants to collapse usings or #regions then we need to compute
+            // synchronously, but only if there are usings or #regions in the file. To
+            // save some work, we'll look for both in a single pass.
+            var collapseRegions = GlobalOptions.GetOption(BlockStructureOptionsStorage.CollapseRegionsWhenFirstOpened, openDocument.Project.Language);
+            var collapseUsings = GlobalOptions.GetOption(BlockStructureOptionsStorage.CollapseImportsWhenFirstOpened, openDocument.Project.Language);
+
+            if (!collapseRegions && !collapseUsings)
+            {
+                return false;
+            }
 
-            // If we contain any #region sections, we want to collapse those automatically on open the first
-            // time a doc is ever opened.  So we need to compute the initial tags synchronously in order to
-            // do that.
-            if (ContainsRegionTag(subjectBuffer.CurrentSnapshot))
+            if (ContainsRegionOrUsing(subjectBuffer.CurrentSnapshot, collapseRegions, collapseUsings, openDocument.Project.Language))
+            {
                 return true;
+            }
 
             return false;
+        }
 
-            static bool ContainsRegionTag(ITextSnapshot textSnapshot)
+        // Internal for testing. Can't use TestAccessor because the base class already defines one
+        internal static bool ContainsRegionOrUsing(ITextSnapshot textSnapshot, bool collapseRegions, bool collapseUsings, string language)
+        {
+            foreach (var line in textSnapshot.Lines)
             {
-                foreach (var line in textSnapshot.Lines)
+                if (collapseRegions && StartsWithRegionTag(line))
+                {
+                    return true;
+                }
+                else if (collapseUsings && IsUsingDeclarationOrExtern(line, language))
                 {
-                    if (StartsWithRegionTag(line))
-                        return true;
+                    return true;
                 }
+            }
+
+            return false;
+
+            static bool StartsWithRegionTag(ITextSnapshotLine line)
+            {
+                if (line.Length < 7)
+                    return false;
+
+                var index = line.Start.Position;
+                Skip(line.Snapshot, ref index, line.End.Position, whitespace: true);
+                // Ignore case here because this deals with VB and C#
+                return line.StartsWith(index, "#region", ignoreCase: true);
```

Comment:
Thanks for the hint. The previous code ignored case here too, I just changed this bit because the previous code allocated, and I'd written the Skip method anyway..